File: codigo/python/runge_kutta.py
"""runge_kutta.py -- Metodo de Runge-Kutta de 4.o orden (RK4). Chapra cap. 25."""


def rk4(f, x0, y0, xf, n):
    h = (xf - x0) / n
    xs, ys = [x0], [y0]
    x, y = x0, y0
    for _ in range(n):
        k1 = f(x, y)
        k2 = f(x + h / 2, y + h / 2 * k1)
        k3 = f(x + h / 2, y + h / 2 * k2)
        k4 = f(x + h,     y + h * k3)
        y += h / 6 * (k1 + 2 * k2 + 2 * k3 + k4)
        x += h
        xs.append(x)
        ys.append(y)
    return xs, ys


if __name__ == "__main__":
    import math
    xs, ys = rk4(lambda x, y: y, 0, 1, 1, 10)
    print(f"y(1) RK4 = {ys[-1]:.8f}   (e = {math.e:.8f})")

File: codigo/matlab/runge_kutta.m
function [x, y] = runge_kutta(f, x0, y0, xf, n)
% RUNGE_KUTTA  Runge-Kutta de 4.o orden (RK4). Chapra cap. 25.
%   Uso:  [x, y] = runge_kutta(@(x,y) y, 0, 1, 1, 10);
    h = (xf - x0)/n;
    x = zeros(1,n+1); y = zeros(1,n+1);
    x(1) = x0; y(1) = y0;
    for i = 1:n
        k1 = f(x(i),     y(i));
        k2 = f(x(i)+h/2, y(i)+h/2*k1);
        k3 = f(x(i)+h/2, y(i)+h/2*k2);
        k4 = f(x(i)+h,   y(i)+h*k3);
        y(i+1) = y(i) + h/6*(k1 + 2*k2 + 2*k3 + k4);
        x(i+1) = x(i) + h;
    end
end
